scripts-available/CDB_CartodbfyTable.sql

```diff
@@ -649,3 +649,837 @@ BEGIN
+--
+-- -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
+--
+--  CDB_CartodbfyTable2(reloid REGCLASS, destschema TEXT DEFAULT NULL)
+--    
+--     Main function, calls the following functions, with a little
+--     logic before the table re-write to avoid re-writing if the table
+--     already has all the necessary columns in place.
+--
+-- (1) _CDB_drop_triggers
+--     As before, this drops all the metadata and geom sync triggers
+--
+-- (2) _CDB_Has_Usable_Primary_ID()
+--     Returns TRUE if it can find a unique integer primary key named
+--    'cartodb_id' or can rename an existing key.
+--     Returns FALSE otherwise.
+--
+-- (3) _CDB_Has_Usable_Geom()
+--     Looks for existing EPSG:4326 and EPSG:3857 geometry columns, and
+--     renames them to the standard names if it can find them, returning TRUE.
+--     If it cannot find both columns in the right EPSG, returns FALSE.
+--
+-- (4) _CDB_Rewrite_Table()
+--     If table does not have a usable primary key and both usable geom
+--     columns it needs to be re-written. Function constructs an appropriate
+--     CREATE TABLE AS SELECT... query and executes it.
+--
+-- (5) _CDB_Add_Indexes()
+--     Checks the primary key column for primary key constraint, adds it if
+--     missing. Check geometry columns for GIST indexes and adds them if missing.
+--
+-- (6) _CDB_create_triggers()
+--     Adds the system metadata and geometry column update triggers back
+--     onto the table.
+--
+-- -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
+
+
+CREATE OR REPLACE FUNCTION _CDB_Columns(OUT pkey TEXT, OUT geomcol TEXT, OUT mercgeomcol TEXT)
+RETURNS record
+AS $$
+BEGIN
+
+pkey := 'cartodb_id';
+geomcol := 'the_geom';
+mercgeomcol := 'the_geom_webmercator';
+
+END;
+$$ LANGUAGE 'plpgsql';
+
+
+CREATE OR REPLACE FUNCTION _CDB_Error(message TEXT, funcname TEXT DEFAULT '_CDB_Error')
+RETURNS void
+AS $$
+BEGIN
+
+  RAISE EXCEPTION 'CDB(%): %', funcname, message;
+
+END;
+$$ LANGUAGE 'plpgsql';
+
+
+CREATE OR REPLACE FUNCTION _CDB_SQL(sql TEXT, funcname TEXT DEFAULT '_CDB_SQL')
+RETURNS void
+AS $$
+BEGIN
+
+  RAISE DEBUG 'CDB(%): %', funcname, sql;
+  EXECUTE sql;
+
+  EXCEPTION
+  WHEN others THEN
+    RAISE EXCEPTION 'CDB(%:%:%): %', funcname, SQLSTATE, SQLERRM, sql;
+
+END;
+$$ LANGUAGE 'plpgsql';
+
+
+-- Find a unique relation name in the given schema, starting from the 
+-- template given. If the template is already unique, just return it; 
+-- otherwise, append an increasing integer until you find a unique variant.
+CREATE OR REPLACE FUNCTION _CDB_Unique_Relation_Name(schemaname TEXT, relationname TEXT)
+RETURNS TEXT
+AS $$
+DECLARE
+  rec RECORD;
+  i INTEGER;
+  newrelname TEXT;
+BEGIN
+
+  i := 0;
+  newrelname := relationname;
+  LOOP
+
+    SELECT c.relname, n.nspname
+    INTO rec
+    FROM pg_class c
+    JOIN pg_namespace n ON c.relnamespace = n.oid
+    WHERE c.relname = newrelname
```

Comment:
Two tricks: finding only names that match the established "pattern" for these things (foo_2) and finding the "next" name. Probably a query that `regexp_matches` the current names can do both at once and spit out an answer suitable for appending the "next" number to. This isn't a particularly deep loop, but replacing loops with SQL is always a good idea, IMO.